src/util/threadpool.h

```diff
@@ -0,0 +1,124 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_THREADPOOL_H
+#define BITCOIN_UTIL_THREADPOOL_H
+
+#include <sync.h>
+#include <util/string.h>
+#include <util/threadnames.h>
+
+#include <condition_variable>
+#include <future>
+#include <queue>
+#include <thread>
+
+class ThreadPool {
+
+private:
+    Mutex cs_work_queue;
+    std::queue<std::function<void()>> m_work_queue GUARDED_BY(cs_work_queue);
+    std::condition_variable m_condition;
+    // Stop indicator
+    std::atomic<bool> m_stop{false};
+    // Worker threads
+    std::vector<std::thread> m_workers;
+
+    void WorkerThread(const std::string& thread_name) EXCLUSIVE_LOCKS_REQUIRED(!cs_work_queue)
+    {
+        util::ThreadRename(std::string{thread_name});
+        WAIT_LOCK(cs_work_queue, wait_lock);
+        while (!m_stop.load()) {
+            std::function<void()> task;
+            {
+                // Wait for the task or until the stop flag is set
+                m_condition.wait(wait_lock,[&]() EXCLUSIVE_LOCKS_REQUIRED(cs_work_queue) { return m_stop.load() || !m_work_queue.empty(); });
+
+                // If stopped, exit worker.
+                if (m_stop.load() && m_work_queue.empty()) {
+                    return;
+                }
+
+                // Pop the task
+                task = std::move(m_work_queue.front());
+                m_work_queue.pop();
+            }
+
+            // Execute the task without the lock
+            WITH_REVERSE_LOCK(wait_lock, task());
+        }
+    }
+
+public:
+    ThreadPool() {}
+
+    ~ThreadPool()
+    {
+        Stop(); // In case it hasn't being stopped.
+    }
+
+    void Start(int num_workers)
+    {
+        if (!m_workers.empty()) throw std::runtime_error("Thread pool already started");
+
+        // Create the workers
+        for (int i=0; i<num_workers; i++) {
+            m_workers.emplace_back(&ThreadPool::WorkerThread, this, "threadpool_worker_" + ToString(i));
+        }
+    }
+
+    void Stop()
+    {
+        // Notify workers and join them.
+        m_stop = true;
+        m_condition.notify_all();
+        for (auto& worker : m_workers) {
+            worker.join();
+        }
+        m_workers.clear();
+        m_stop = false;
+    }
+
+    template<class T> EXCLUSIVE_LOCKS_REQUIRED(!cs_work_queue)
+    auto Submit(T task) -> std::future<decltype(task())>
+    {
+        auto ptr_task = std::make_shared<std::packaged_task<decltype(task()) ()>>(std::move(task));
+        std::future<decltype(task())> future = ptr_task->get_future();
+        {
+            LOCK(cs_work_queue);
+            m_work_queue.emplace([=]() {
+                (*ptr_task)();
+            });
+        }
+        m_condition.notify_one();
+        return future;
+    }
+
+    // Synchronous processing
+    void ProcessTask() EXCLUSIVE_LOCKS_REQUIRED(!cs_work_queue)
```

Comment:
This would make sense if we wanted to reuse this for CCheckQueue. Then, we could just loop and `ProcessTask` on the main thread once when we call `Wait`.